bluesky_queueserver/manager/comms.py

```diff
@@ -412,3 +414,211 @@ def _pipe_receive(self):
+            is passed to the server.
+
+        Returns
+        -------
+        dict
+            Message returned by the server.
+
+        Raises
+        ------
+        CommTimeoutError
+            Raised if communication error occurs and ``raise_timeout_exceptions`` is set ``True``.
+        """
+
+        # Send empty dictionary if no parameters are passed
+        params = params or {}
+
+        async with self._lock_zmq:
+            try:
+                msg_out = self._create_msg(method=method, params=params)
+                msg_in = await self._zmq_communicate(msg_out)
+            except Exception as ex:
+                # This is very likely a timeout (RE Manager is not responding)
+                self._zmq_socket_restart()
+                errmsg = f"ZMQ communication error: {str(ex)}"
+                if self._raise_timeout_exceptions:
+                    raise CommTimeoutError(errmsg)
+                msg_in = {"success": False, "msg": errmsg}
+            return msg_in
+
+    def close(self):
+        """
+        Close ZMQ socket. Call to close socket if the object is no longer needed, but may
+        not be destroyed for some time.
+        """
+        self._zmq_socket.close()
+
+
+def zmq_single_request(method, params=None, *, zmq_server_address=None):
+    """
+    Send a single request to ZMQ server. The function opens the socket, sends
+    a single ZMQ request and closes the socket. The function is not expected
+    to raise exceptions. In case of communication error the return value
+    of ``msg`` is ``None`` and ``err_msg`` contains the error message. Otherwise
+    ``err_msg`` is empty and ``msg`` contains the dictionary returned by the server.
+
+    Parameters
+    ----------
+    method: str
+        Name of the method called in RE Manager
+    params: dict or None
+        Dictionary of parameters (payload of the message). If ``None`` then
+        the message is sent with empty payload: ``params = {}``.
+
+    Returns
+    -------
+    msg: dict or None
+        Message received from RE Manager in response to the request. None if communication
+        error (timeout) occurred.
+    err_msg: str
+        Contains a message in case communication error (timeout) occurs. Empty string otherwise.
+    """
+
+    msg_received = None
+
+    async def send_request(method, params):
+        nonlocal msg_received
+        zmq_to_manager = ZMQCommSendAsync(zmq_server_address=zmq_server_address)
+        msg_received = await zmq_to_manager.send_message(method=method, params=params)
+        zmq_to_manager.close()
+
+    try:
+        asyncio.run(send_request(method, params))
+
+        msg = msg_received
+        msg_err = ""
+    except Exception as ex:
+        msg = None
+        msg_err = str(ex)
+
+    if msg_err:
+        logger.warning("Communication with RE Manager failed: %s", str(msg_err))
+
+    return msg, msg_err
```

Comment:
Why not raise this? This pattern is common in JS but strikes me as unusual in Python.

Reply:
There is no particular reason except that this function is a product of evolution and it was moved here almost without change. I think I will fix it when I make another pass over this code (this function is widely used in tests, so there is some editing involved). Also this function is odd, because it does a lot of unnecessary stuff just to send a single zmq request, so it should be used in production only in cases where requests are infrequent or single (as in the CLI tool). We may need to think about another API for use in event-driven applications (such as Qt applications) that would provide modes including blocking wait, polling and callbacks. I think that if we cover both synchronous or thread-based and asynchronous cases, there is better chance the API will be used.